Remove `AbstractSeparator` and refactor descendants

ramchale opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Refactor;
- DashToCamelCase
- SeparatorToCamelCase
- UnderscoreToCamelCase

Remove;
- AbstractSeparator

ramchale commented on docs/book/v3/migration/v2-to-v3.md:
```diff
@@ -251,6 +255,11 @@ Support for these formats has been removed so the following classes are no longe
 
 The following compression formats are still available: `Bz2`, `Gz`, `Tar` and `Zip`
 
+### Removal of the `AbstractSeparator` class
+
+Various filters such as `SeparatorToCamelCase` inherited from the abstract class `AbstractSeparator` to reuse the construction and accessor methods.
+The classes that inherited now use composition to continue provision the expected behaviour, without the risk of changes to the abstract class unintentionally modifying the interface of inheriting classes.
+
```
I've had a go at this, but would appreciate some input from someone with more context about the future direction of this package

gsteel replied:
Perhaps just drop the docs changes here - I'll create a TODO issue for the migration guide so that I remember to cover all the relevant topics before 3.0.0 is released 👍

ramchale replied:
Cool. Thanks 
I've taken that out

gsteel commented on src/Word/SeparatorToCamelCase.php:
```diff
@@ -4,65 +4,58 @@
  *     separator?: string,
- *     ...
  * }
  * @template TOptions of Options
```
This template is no longer necessary

ramchale replied:
Thanks for taking a look. I've amended that

gsteel reviewed (changes requested):
This is looking good to me @ramchale - Just a minor tweak and an updated baseline and it should be good to go 👍

ramchale committed:
Remove AbstractSeparator

Refactor;
- DashToCamelCase
- SeparatorToCamelCase
- UnderscoreToCamelCase

Remove;
- AbstractSeparator

Signed-off-by: ramchale <legionsofbob@gmail.com>

gsteel reviewed (approved):
LGTM! Thank you @ramchale 💪

froschdesign commented:
@gsteel 
I added to this pull request to #177.
